components/server/src/code-sync/code-sync-service.ts

```diff
@@ -0,0 +1,214 @@
+/**
+ * Copyright (c) 2020 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+import { status } from 'grpc';
+import fetch from "node-fetch";
+import { User } from '@gitpod/gitpod-protocol/lib/protocol';
+import bodyParser = require('body-parser');
+import * as util from 'util';
+import * as express from 'express';
+import { inject, injectable } from 'inversify';
+import { BearerAuth } from '../auth/bearer-authenticator';
+import { isWithFunctionAccessGuard } from '../auth/function-access';
+import { CodeSyncResourceDB } from '@gitpod/gitpod-db/lib/typeorm/code-sync-resource-db';
+import { ALL_SERVER_RESOURCES, ServerResource } from '@gitpod/gitpod-db/lib/typeorm/entity/db-code-sync-resource';
+import { BlobServiceClient } from '@gitpod/content-service/lib/blobs_grpc_pb';
+import { DeleteRequest, DownloadUrlRequest, DownloadUrlResponse, UploadUrlRequest, UploadUrlResponse } from '@gitpod/content-service/lib/blobs_pb';
+import { log } from '@gitpod/gitpod-protocol/lib/util/logging';
+import uuid = require('uuid');
+
+// By default: 6 kind of resources * 20 revs * 1Mb = 120Mb max in the content service for user data.
+const defautltRevLimit = 20;
+// It should keep it aligned with client_max_body_size for /code-sync location.
+const defaultContentLimit = '1Mb';
+const codeSyncConfig: Partial<{
+    revLimit: number
+    contentLimit: number
+    resources: {
+        [resource: string]: {
+            revLimit?: number
+        }
+    }
+}> = JSON.parse(process.env.CODE_SYNC_CONFIG || "{}");
+
+const objectPrefix = 'code-sync/';
+function toObjectName(resource: ServerResource, rev: string): string {
+    return objectPrefix + resource + '/' + rev;
+}
+
+@injectable()
+export class CodeSyncService {
+
+    @inject(BearerAuth)
+    private readonly auth: BearerAuth;
+
+    @inject(BlobServiceClient)
+    private readonly blobs: BlobServiceClient;
+
+    @inject(CodeSyncResourceDB)
+    private readonly db: CodeSyncResourceDB;
+
+    get apiRouter(): express.Router {
+        const router = express.Router();
+        router.use((_, res, next) => {
+            // to correlate errors reported by users with errors logged by the server
+            res.setHeader('x-operation-id', uuid.v4());
```

Comment:
Could you elaborate a bit on how this correlation works?
Also, could we re-use one of our usual correlation IDs here somehow (OWI)?

Reply:
If the request fails a user will get a notification with this id included which he can use to open the issue. On the server side failed request is also logged with such header to match them.

OWI stands for? if (owner, workspace, isntance) then it wont't unique identify a request to the code sync storage.